File: packages/sdk/go/context.go
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// CaracalContext: identity and delegation context threaded through context.Context.

package sdk

import (
	"context"
	"errors"
)

type contextKey struct{}

// CaracalContext carries the Caracal identity and delegation state
// for a single execution path.
type CaracalContext struct {
	SubjectToken     string
	ZoneID           string
	ClientID         string
	AgentSessionID   string
	DelegationEdgeID string
	ParentEdgeID     string
	SessionID        string
	TraceID          string
	Hop              int
}

// AuthoritySummary is a redacted operator view of the bound authority chain.
type AuthoritySummary struct {
	ZoneID                      string
	ApplicationID               string
	AuthoritySessionID          string
	AgentRunID                  string
	DelegatedPermissionID       string
	ParentDelegatedPermissionID string
	TraceID                     string
	Hop                         int
	Chain                       []string
}

// Bind returns a new context.Context carrying c.
func Bind(parent context.Context, c CaracalContext) context.Context {
	return context.WithValue(parent, contextKey{}, c)
}

// Current returns the CaracalContext bound on ctx and whether one was found.
func Current(ctx context.Context) (CaracalContext, bool) {
	v := ctx.Value(contextKey{})
	if v == nil {
		return CaracalContext{}, false
	}
	return v.(CaracalContext), true
}

// Capture returns a copy of the current CaracalContext for explicit task boundaries.
func Capture(ctx context.Context) (CaracalContext, bool) {
	return Current(ctx)
}

// FromEnvelope builds a CaracalContext from a deserialized Envelope.
func FromEnvelope(env Envelope, zoneID, clientID string) (CaracalContext, error) {
	if env.SubjectToken == "" {
		return CaracalContext{}, errors.New("caracal: envelope missing subject token")
	}
	return CaracalContext{
		SubjectToken:     env.SubjectToken,
		ZoneID:           zoneID,
		ClientID:         clientID,
		AgentSessionID:   env.AgentSessionID,
		DelegationEdgeID: env.DelegationEdgeID,
		ParentEdgeID:     env.ParentEdgeID,
		SessionID:        env.SessionID,
		TraceID:          env.TraceID,
		Hop:              env.Hop,
	}, nil
}

// ToEnvelope projects a CaracalContext to a wire Envelope.
func ToEnvelope(c CaracalContext) Envelope {
	return Envelope{
		SubjectToken:     c.SubjectToken,
		AgentSessionID:   c.AgentSessionID,
		DelegationEdgeID: c.DelegationEdgeID,
		ParentEdgeID:     c.ParentEdgeID,
		SessionID:        c.SessionID,
		TraceID:          c.TraceID,
		Hop:              c.Hop,
	}
}

// DescribeAuthority returns a redacted authority-chain summary for logs and diagnostics.
func DescribeAuthority(ctx context.Context) (AuthoritySummary, bool) {
	c, ok := Current(ctx)
	if !ok {
		return AuthoritySummary{}, false
	}
	return DescribeCaracalContext(c), true
}

// DescribeCaracalContext projects a CaracalContext into user-facing authority terms.
func DescribeCaracalContext(c CaracalContext) AuthoritySummary {
	chain := []string{}
	if c.SessionID != "" {
		chain = append(chain, "authority:"+c.SessionID)
	}
	if c.AgentSessionID != "" {
		chain = append(chain, "agent-run:"+c.AgentSessionID)
	}
	if c.ParentEdgeID != "" {
		chain = append(chain, "parent-delegated-permission:"+c.ParentEdgeID)
	}
	if c.DelegationEdgeID != "" {
		chain = append(chain, "delegated-permission:"+c.DelegationEdgeID)
	}
	return AuthoritySummary{
		ZoneID:                      c.ZoneID,
		ApplicationID:               c.ClientID,
		AuthoritySessionID:          c.SessionID,
		AgentRunID:                  c.AgentSessionID,
		DelegatedPermissionID:       c.DelegationEdgeID,
		ParentDelegatedPermissionID: c.ParentEdgeID,
		TraceID:                     c.TraceID,
		Hop:                         c.Hop,
		Chain:                       chain,
	}
}

File: packages/sdk/ts/src/context.ts
/*
 * Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
 * Caracal, a product of Garudex Labs
 *
 * CaracalContext: bound identity and delegation context propagated across async boundaries.
 */

import { AsyncLocalStorage } from 'node:async_hooks'
import { Envelope } from './envelope.js'

export interface CaracalContext {
  subjectToken: string
  zoneId: string
  applicationId: string
  agentSessionId?: string
  delegationEdgeId?: string
  parentEdgeId?: string
  sessionId?: string
  traceId?: string
  hop: number
}

export interface AuthoritySummary {
  zoneId: string
  applicationId: string
  sessionId?: string
  agentSessionId?: string
  delegationEdgeId?: string
  parentEdgeId?: string
  traceId?: string
  hop: number
  chain: string[]
}

const storage = new AsyncLocalStorage<CaracalContext>()

export function current(): CaracalContext | undefined {
  return storage.getStore()
}

export function captureContext(): CaracalContext | undefined {
  const ctx = current()
  return ctx ? { ...ctx } : undefined
}

export function bind<T>(ctx: CaracalContext, fn: () => T | Promise<T>): T | Promise<T> {
  return storage.run(ctx, fn)
}

export function withOverrides<T>(patch: Partial<CaracalContext>, fn: () => T | Promise<T>): T | Promise<T> {
  const base = current()
  if (!base) throw new Error('withOverrides requires an existing Caracal context')
  return storage.run({ ...base, ...patch }, fn)
}

export function toEnvelope(ctx: CaracalContext): Envelope {
  return {
    subjectToken: ctx.subjectToken,
    agentSessionId: ctx.agentSessionId,
    delegationEdgeId: ctx.delegationEdgeId,
    parentEdgeId: ctx.parentEdgeId,
    sessionId: ctx.sessionId,
    traceId: ctx.traceId,
    hop: ctx.hop,
  }
}

export function fromEnvelope(env: Envelope, base: { zoneId: string; applicationId: string }): CaracalContext {
  if (!env.subjectToken) throw new Error('envelope missing subject token')
  return {
    subjectToken: env.subjectToken,
    zoneId: base.zoneId,
    applicationId: base.applicationId,
    agentSessionId: env.agentSessionId,
    delegationEdgeId: env.delegationEdgeId,
    parentEdgeId: env.parentEdgeId,
    sessionId: env.sessionId,
    traceId: env.traceId,
    hop: env.hop,
  }
}

export function describeAuthority(ctx: CaracalContext | undefined = current()): AuthoritySummary | undefined {
  if (!ctx) return undefined
  const chain: string[] = []
  if (ctx.sessionId) chain.push(`session:${ctx.sessionId}`)
  if (ctx.agentSessionId) chain.push(`agent-session:${ctx.agentSessionId}`)
  if (ctx.parentEdgeId) chain.push(`parent-edge:${ctx.parentEdgeId}`)
  if (ctx.delegationEdgeId) chain.push(`delegation-edge:${ctx.delegationEdgeId}`)
  return {
    zoneId: ctx.zoneId,
    applicationId: ctx.applicationId,
    sessionId: ctx.sessionId,
    agentSessionId: ctx.agentSessionId,
    delegationEdgeId: ctx.delegationEdgeId,
    parentEdgeId: ctx.parentEdgeId,
    traceId: ctx.traceId,
    hop: ctx.hop,
    chain,
  }
}
